vpr/src/route/connection_box_lookahead_map.cpp

```diff
@@ -21,370 +22,386 @@
-    } else {
-        VPR_FATAL_ERROR(VPR_ERROR_ROUTE, "Unknown channel type %d", rr_type);
+    // store entries into the matrices
+    for (const auto& entry : costs) {
+        int seg = entry.first.seg_index;
+        int box = size_t(entry.first.box_id);
+        const auto& seg_box_bounds = bounds[seg][box];
+        int x = entry.first.delta.x() - seg_box_bounds.xmin();
+        int y = entry.first.delta.y() - seg_box_bounds.ymin();
+        cost_map_[seg][box][x][y] = entry.second.cost_entry;
     }
-}
 
-class StartNode {
-  public:
-    StartNode(int start_x, int start_y, t_rr_type rr_type, int seg_index)
-        : start_x_(start_x)
-        , start_y_(start_y)
-        , rr_type_(rr_type)
-        , seg_index_(seg_index)
-        , index_(0) {}
-    int get_next_node() {
-        const auto& device_ctx = g_vpr_ctx.device();
-        const std::vector<int>& channel_node_list = get_rr_node_indcies(
-            rr_type_, start_x_, start_y_);
-
-        for (; index_ < channel_node_list.size(); index_++) {
-            int node_ind = channel_node_list[index_];
-
-            if (node_ind == OPEN || device_ctx.rr_nodes[node_ind].capacity() == 0) {
+    // fill the holes
+    for (size_t seg = 0; seg < seg_count_; seg++) {
+        for (size_t box = 0; box < box_count_; box++) {
+            const auto& seg_box_bounds = bounds[seg][box];
+            if (seg_box_bounds.empty()) {
                 continue;
             }
+            auto& matrix = cost_map_[seg][box];
+
+            // calculate delay penalty
+            float min_delay = 0.f, max_delay = 0.f;
+            vtr::Point<int> min_location(0, 0), max_location(0, 0);
+            for (unsigned ix = 0; ix < matrix.dim_size(0); ix++) {
+                for (unsigned iy = 0; iy < matrix.dim_size(1); iy++) {
+                    Cost_Entry& cost_entry = matrix[ix][iy];
+                    if (cost_entry.valid()) {
+                        if (cost_entry.delay < min_delay) {
+                            min_delay = cost_entry.delay;
+                            min_location = vtr::Point<int>(ix, iy);
+                        }
+                        if (cost_entry.delay > max_delay) {
+                            max_delay = cost_entry.delay;
+                            max_location = vtr::Point<int>(ix, iy);
+                        }
+                    }
+                }
+            }
+            float delay_penalty = (max_delay - min_delay) / static_cast<float>(std::max(1, manhattan_distance(max_location, min_location)));
+
+            // find missing cost entries and fill them in by copying a nearby cost entry
+            std::vector<std::tuple<unsigned, unsigned, Cost_Entry>> missing;
+            bool couldnt_fill = false;
+            auto shifted_bounds = vtr::Rect<int>(0, 0, seg_box_bounds.width(), seg_box_bounds.height());
+            int max_fill = 0;
+            for (unsigned ix = 0; ix < matrix.dim_size(0) && !couldnt_fill; ix++) {
```

Comment:
I'm not sure we should stop filling on x just because one fill failed?

Reply:
Fixed.